function-images/tests/video-bench/decoder/decode.py

```diff
@@ -0,0 +1,112 @@
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from __future__ import print_function
+
+import sys
+import os
+# adding python tracing sources to the system path
+sys.path.insert(0, os.getcwd() + '/../proto/')
+sys.path.insert(0, os.getcwd() + '/../../../../utils/tracing/python')
+import tracing
+import videoservice_pb2_grpc
+import videoservice_pb2
+
+import io
+import cv2
+import grpc
+import tempfile
+import argparse
+import asyncio
+
+
+from concurrent import futures
+from timeit import default_timer as now
+
+# USE ENV VAR "DecoderFrames" to set the number of frames to be sent
+parser = argparse.ArgumentParser()
+parser.add_argument("-addr", "--addr", dest = "addr", default = "recog.default.192.168.1.240.sslip.io:80", help="recog address")
+parser.add_argument("-sp", "--sp", dest = "sp", default = "80", help="serve port")
+parser.add_argument("-frames", "--frames", dest = "frames", default = "1", help="Default number of frames- overwritten by environment variable")
+parser.add_argument("-zipkin", "--zipkin", dest = "url", default = "http://zipkin.istio-system.svc.cluster.local:9411/api/v2/spans", help="Zipkin endpoint url")
+
+args = parser.parse_args()
+
+if tracing.IsTracingEnabled():
+    tracing.initTracer("decoder", url=args.url)
+    tracing.grpcInstrumentClient()
+    tracing.grpcInstrumentServer()
+
+def decode(bytes):
+    with tracing.Span("Making Tempfile"):
+        start_tempfile = now()
+        temp = tempfile.NamedTemporaryFile(suffix=".mp4")
+        temp.write(bytes)
+        temp.seek(0)
+        end_tempfile_write = now()
+        write_time = (end_tempfile_write - start_tempfile)*1000
+        print("tempfile write time: %dms" %write_time)
+
+    all_frames = [] 
+    with tracing.Span("Get frames"):
+        vidcap = cv2.VideoCapture(temp.name)
+        for i in range(os.getenv('DecoderFrames', int(args.frames))):
+            success,image = vidcap.read()
+            all_frames.append(cv2.imencode('.jpg', image)[1].tobytes())
+
+    return all_frames
+
+
+
+class VideoDecoderServicer(videoservice_pb2_grpc.VideoDecoderServicer):
+    def Decode(self, request, context):
+        out = decode(request.video)
+        with tracing.Span("Send all frames"):  
+            for i in range(os.getenv('DecoderFrames', int(args.frames))):
+                result_future = self.Recognise(out[i])
+                print("got a result")
+        # just returns the final result
+        print(result_future.result().classification)
```

Comment:
ok, this is functionally wrong: you need to wait for all recog calls to return but you only wait for the last one in my understanding.

You should wait for all calls to return and the only then close the scope of the tracing span.